docs/codeql/codeql-language-guides/customizing-library-models-for-go.rst

```diff
@@ -0,0 +1,338 @@
+
+Example: Add flow through the ``Join`` function
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This example shows how the Go query pack models flow through a method for a simple case.
+This pattern covers many of the cases where we need to summarize flow through a function that is stored in a library or framework outside the repository.
+
+.. code-block:: go
+
+    func TaintFlow() {
+        elems := []string{"Hello", "World"}
+        sep := " "
+        t := strings.Join(elems, sep) // There is taint flow from elems and sep to t.
+        ...
+    }
+
+We need to add tuples to the ``summaryModel``\(package, type, subtypes, name, signature, ext, input, output, kind, provenance) extensible predicate by updating a data extension file:
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/go-all
+         extensible: summaryModel
+       data:
+         - ["strings", "", False, "Join", "", "", "Argument[0]", "ReturnValue", "taint", "manual"]
+         - ["strings", "", False, "Join", "", "", "Argument[1]", "ReturnValue", "taint", "manual"]
+
+Since we are adding flow through a method, we need to add tuples to the ``summaryModel`` extensible predicate.
+Each tuple defines flow from one argument to the return value.
+The first row defines flow from the first argument (``elems`` in the example) to the return value (``t`` in the example) and the second row defines flow from the second argument (``sep`` in the example) to the return value (``t`` in the example).
+
+The first five values identify the function to be modeled as a summary.
+These are the same for both of the rows above as we are adding two summaries for the same method.
+
+- The first value ``strings`` is the package name.
+- The second value ``""`` is left blank, since the function is not a method of a type.
+- The third value ``False`` is a flag that indicates whether or not the summary also applies to all overrides of the method.
+- The fourth value ``Join`` is the function name.
+- The fifth value ``""`` is left blank, since specifying the signature is optional and Go does not allow multiple signature overloads for the same function.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the summary.
+
+- The seventh value is the access path to the input (where data flows from). ``Argument[0]`` is the access path to the first argument (``elems`` in the example) and ``Argument[1]`` is the access path to the second argument (``sep`` in the example).
+- The eighth value ``ReturnValue`` is the access path to the output (where data flows to), in this case ``ReturnValue``, which means that the input flows to the return value.
+- The ninth value ``taint`` is the kind of the flow. ``taint`` means that taint is propagated through the call.
```

Comment:
Mention that this could be `value` and the difference between them?

Reply:
Would examples relating to the `slices` package make sense? Something like `slices.MaxFunc` is comparable to the Java example (`Stream#map`) in terms of:

1. Value flow `element` -> `func`
2. Value flow `element` -> `return value`

Though that brings up some other questions:

- How does MaD-for-Go handle generics?
- Are elements of slices handled by `.ArrayElement`? (Based on how `append` is modeled I think so, but want to check)

Reply:
I'm thinking something like:

for `func MaxFunc[S ~[]E, E any](x S, cmp func (a, b E) int) E`

```yaml
- ["slices", "", True, "MaxFunc", "", "", "Argument[0].ArrayElement", "Argument[1].Parameter[0]", "value", "manual"] 
- ["slices", "", True, "MaxFunc", "", "", "Argument[0].ArrayElement", "Argument[1].Parameter[1]", "value", "manual"] 
- ["slices", "", True, "MaxFunc", "", "", "Argument[0].ArrayElement", "ReturnValue", "value", "manual"] 
```

Reply:
Other decent options I can think of:

- `slices.Reverse` (`Argument[0].ArrayElement -> ReturnValue.ArrayElement`)
- `slices.Concat`, though that has some annoyances around varargs

Reply:
The go dataflow library doesn't currently deal with lambdas, so I don't think the first two of those models would work. `slices.Concat` might be good - yes, the variadic argument makes it more complicated, but it also gives you an opportunity to explain how to model variadic arguments, which is currently missing. (Not that these docs have to be exhaustive.)